R/src_gen.R

```diff
@@ -0,0 +1,207 @@
+#' @title Generate Synthetic Rating Curve
+#' @param comids A set of COMIDs. See \link{find_comids}.
+#' @param hand A HAND \code{raster}.
+#' @param stage A set of stage values. Defaults to \code{0:20}.
+#' @param progress If \code{TRUE}, output a progress bar.
+#' @return A \code{data.frame} representing the SRC table.
+#' @importFrom nhdplusTools get_vaa
+#' @importFrom data.table :=
+#' @importFrom plyr .
+#' @export
+get_src <- function(comids, hand, stage = 0:20, progress = TRUE) {
+    if (progress) {
+        pb <- progress::progress_bar$new(
+            total = length(stage) + 3,
+            format = " generating SRC [:bar] :percent eta: :eta"
+        )
+
+        pb$tick(0)
+    }
+
+    if (missing(comids) & !missing(hand)) {
+        hand_bb <- sf::st_bbox(hand) %>%
+                   sf::st_as_sfc() %>%
+                   sf::st_as_sf()
+
+        comids <- find_comids(hand_bb)
+    }
+
+    # If the HAND raster is not passed, get from AWS
+    if (missing(hand) & !missing(comids)) {
+        nhd <- nhdplusTools::get_nhdplus(comid = comids)
+        hand <- get_hand_raster(nhd)
+    }
+
+    # Fix NA values
+    hand[is.na(hand[])] <- 0
+    # Generate Slope Raster
+    slope <- raster::terrain(hand)
+    slope[is.na(slope[])] <- 0
+    # Get raster cell resolution
+    res <- compute_cellres(hand)
+    # Create HAND_Slope data frame
+
+    if (progress) pb$tick()
+
+    hand_slope <-
+        data.frame(
+            hand = raster::getValues(hand),
+            slope = raster::getValues(slope)
+        ) %>%
+        na.omit() %>%
+        dplyr::arrange(hand) %>%
+        dplyr::mutate(
+            res = res,
+            BA  = res * res * sqrt(1 + slope)
+        )
+
+    if (progress) pb$tick()
+
+    # Get VAAs via nhdplusTools
+    vaa <-
+        nhdplusTools::get_vaa(
+            atts = c(
+                "pathlength",
+                "arbolatesu",
+                "lengthkm",
+                "areasqkm",
+                "slope",
+                "roughness"
+            )
+        ) %>%
+        dplyr::filter(comid %in% comids)
+
+    if (progress) pb$tick()
+
+    synthetic_rating_curve <-
+        lapply(
+            stage,
+            FUN = function(y) {
+
+                if (progress) pb$tick()
+
+                vaa %>%
+                    na.omit() %>%
+                    dplyr::group_by(comid) %>%
```

Comment:
Isnt the group by whats slow? Was it helped by data.table?

Reply:
The first `group_by` is not slow, and gets completed in roughly ~1-3 seconds for 400+ COMIDs. The grouped operation involving `compute_flat_tub` is the bottleneck, and makes the operation slow for a list of 25 or more COMIDs, see [the table here](https://github.com/program--/FloodMapping/blob/8a65849899b351a4ac71d9594691c35cb6362a6f/vignettes/generateSRC.Rmd#L58-L65).

I think that `data.table` definitely helped, but the operation efficiency decreases exponentially (or similar to exponentially) as both the AOI and list of COMIDs grows larger. I'm not currently sure how to reconcile the operation, as the grouping is required to ensure Q isn't set to the largest value returned for all the COMIDs. (this is what I found out happens without grouping)

One way that might work, but also might still slow down the function, is to use `dplyr::rowise()` instead of grouping. I'd have to run benchmarks to see if it is any faster though.